### ISSUE TITLE
src/lua-utils.cpp:154:5: error: ‘luaL_traceback’ was not declared in this scope

### ISSUE THREAD

**Issue by gen2** [openstreetmap/osm2pgsql#1151]
Can't compile osm2pgsql anymore:
...
```
[35/42] /usr/bin/x86_64-pc-linux-gnu-g++ -DDEFAULT_STYLE=\"/usr/share/osm2pgsql/default.style\" -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -I/mnt/gentoo/tmp/portage/sci-
geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/src -I. -isystem /mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/contrib/libosm
ium/include -isystem /mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/contrib/protozero/include -isystem /mnt/gentoo/tmp/portage/sci-ge
osciences/osm2pgsql-99999999/work/osm2pgsql-99999999/contrib/fmt/include -isystem /usr/include/postgresql  -DACCEPT_USE_OF_DEPRECATED_PROJ_API_H=1 -march=sandybridge -fo
mit-frame-pointer -O2 -pipe   -Wall -std=c++11 -MD -MT src/CMakeFiles/osm2pgsql_lib.dir/lua-utils.cpp.o -MF src/CMakeFiles/osm2pgsql_lib.dir/lua-utils.cpp.o.d -o src/CMa
keFiles/osm2pgsql_lib.dir/lua-utils.cpp.o -c /mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/src/lua-utils.cpp
FAILED: src/CMakeFiles/osm2pgsql_lib.dir/lua-utils.cpp.o
/usr/bin/x86_64-pc-linux-gnu-g++ -DDEFAULT_STYLE=\"/usr/share/osm2pgsql/default.style\" -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -I/mnt/gentoo/tmp/portage/sci-geoscien
ces/osm2pgsql-99999999/work/osm2pgsql-99999999/src -I. -isystem /mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/contrib/libosmium/incl
ude -isystem /mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/contrib/protozero/include -isystem /mnt/gentoo/tmp/portage/sci-geoscience
s/osm2pgsql-99999999/work/osm2pgsql-99999999/contrib/fmt/include -isystem /usr/include/postgresql  -DACCEPT_USE_OF_DEPRECATED_PROJ_API_H=1 -march=sandybridge -fomit-fram
e-pointer -O2 -pipe   -Wall -std=c++11 -MD -MT src/CMakeFiles/osm2pgsql_lib.dir/lua-utils.cpp.o -MF src/CMakeFiles/osm2pgsql_lib.dir/lua-utils.cpp.o.d -o src/CMakeFiles/
osm2pgsql_lib.dir/lua-utils.cpp.o -c /mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/src/lua-utils.cpp
/mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/src/lua-utils.cpp: In function ‘int pcall_error_traceback_handler(lua_State*)’:
/mnt/gentoo/tmp/portage/sci-geosciences/osm2pgsql-99999999/work/osm2pgsql-99999999/src/lua-utils.cpp:154:5: error: ‘luaL_traceback’ was not declared in this scope
  154 |     luaL_traceback(lua_state, lua_state, msg, 1); // append a standard traceback
      |     ^~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
 * ERROR: sci-geosciences/osm2pgsql-99999999::gentoo failed (compile phase):
 *   ninja -v -j1 -l0 failed
```
What do i miss here? Any help appreciated!

**Comment by mmd-osm** [comment]
Can you maybe ask in your Gentoo community why they don’t support luaL_traceback?

**Comment by gen2** [comment]
> Can you maybe ask in your Gentoo community why they don’t support luaL_traceback?

Is it part of lua 5.1.x?

**Comment by mmd-osm** [comment]
It works on Lua 5.1.4 (LuaJIT 2.1.0-beta3), possibly they have backported some features from 5.2? Can you try with 5.2 or LuaJIT instead?

**Comment by gen2** [comment]
Am May 3, 2020 9:22:07 PM UTC schrieb mmd <notifications@github.com>:
> It works on Lua 5.1.4 (LuaJIT 2.1.0-beta3), possibly they have
> backported some features from 5.2? Can you try with 5.2 or LuaJIT
> instead?

I've lua 5.1.5 installed. Everything newer is masked on Gentoo.

Is there a way to disable osm2pgswl's lua extension? I simply don't need it.

-- 
sent from my phone

**Comment by mmd-osm** [comment]
Just use a released version (rather than the master branch), or wait for this to be fixed.

**Comment by gen2** [comment]
> Can you maybe ask in your Gentoo community why they don’t support luaL_traceback?

is reported: https://bugs.gentoo.org/720916
Let's see what their decision is. Maybe they ask for dev-lang/lua:5.2 as (build?) dependency into the ebuild.

**Comment by mmd-osm** [comment]
Gentoo ebuild dependencies are not managed in this repository. Whatever their decision is, it has no immediate impact on this repo.

**Comment by gen2** [comment]
> Gentoo ebuild dependencies are not managed in this repository. Whatever their decision is, it has no immediate impact on this repo. THis is something that needs to be handled by someone on Gentoo side.

Of course, and i know this. But since they're experts in system building, i'll follow their advice first.

**Comment by gen2** [comment]
> It works on Lua 5.1.4 (LuaJIT 2.1.0-beta3), possibly they have backported some features from 5.2. Can you try with 5.2 or LuaJIT instead?

i tried luajit w/ lua52compat enabled, but since their header files are in /usr/include/luajit-2.0 it still failes. But at least luaL_traceback is supported:

```
# fgrep -rn luaL_traceback /usr/include/luajit-2.0/
/usr/include/luajit-2.0/lauxlib.h:85:LUALIB_API void luaL_traceback (lua_State *L, lua_State *L1, const char *msg,
```


> Seems like we need to add an #ifdef in luaX_pcall and fallback to the old behaviour if Lua is too old.
> 
> http://lua.2524044.n2.nabble.com/Traceback-from-C-td7655196.html has some hints how to implement this feature for 5.1. As this isn't part of our unit tests, I'm very reluctant to implement it here, and would rather disable stack traces for 5.1. Also, our installation instructions assume Lua 5.2 everywhere (or LuaJIT).


What about a checking for sufficient lua version (and throwing an error) before starting the compiler?

**Comment by mmd-osm** [comment]
We already have a pull request #1153 in place which will disable stack traces for 5.1. Once that is merged, it shoudl solve your issue.

Re the different location for LuaJIT on Gentoo, that would be another issue, then. We don't really have Gentoo in our unit tests, and I've read that Lua is massively broken in Gentoo at times. I guess you're a bit on your own here.

**Comment by gen2** [comment]
> We already have a pull request #1153 in place which will disable stack traces for 5.1. Once that is merged, it shoudl solve your issue.

I just checked it. But do you know why the current master doesn't use the installed luajit, but sticks to lua5.1?


> Re the different location for LuaJIT on Gentoo, that would be another issue, then. We don't really have Gentoo in our unit tests, and I've read that Lua is massively broken in Gentoo at times. I guess you're a bit on your own here.

Where did you read this? I've no problems so far, but i'm not a heavy lua user anyway.



> I also wonder why Gentoo is pulling from the master branch. Usually, we recommend released versions only, as it's really messy to have 10 people with 10 different osm2pgsql versions running. Maybe it would be good to have some guidance in the README file on this topic.

The other currently available version is  0.96.0, and that's not sufficient for me, so i've unmasked this git  master branch version on our servers. I'll look into a version based ebuild approach by using osm2pgsql's tags. It works quite well w/ [mapnik](https://github.com/mapnik/mapnik),but they also provide [tar.gz](https://mapnik.org/pages/downloads.html) files.

**Comment by mmd-osm** [comment]
> But do you know why the current master doesn't use the installed luajit, but sticks to lua5.1?

Again, this is a topic for another issue. Let's not mix different issues here.

> The other currently available version is 0.96.0

1.2.1 is the latest release: https://github.com/openstreetmap/osm2pgsql/releases

**Comment by gen2** [comment]
> > But do you know why the current master doesn't use the installed luajit, but sticks to lua5.1?
> 
> Again, this is a topic for another issue. Let's not mix different issues here.

here we go:  #1154.

**Comment by gen2** [comment]
@mmd-osm & @lonvia : tnx a lot for the quick fix!
